src/main/java/com/minecolonies/coremod/compatibility/jei/transer/PrivateCraftingTeachingTransferHandler.java

```diff
@@ -79,20 +94,37 @@ public IRecipeTransferError transferRecipe(
                 inputIndex++;
             }
         }
-
-        final InventoryCrafting craftMatrix = new InventoryCrafting(new Container() {
+        final int size = craftingGUIBuilding.isComplete() ? 3 : 2;
+        final InventoryCrafting craftMatrix = new InventoryCrafting(new Container()
+        {
             @Override
             public boolean canInteractWith(final EntityPlayer entityPlayer)
             {
                 return false;
             }
-        }, 2, 2);
+        }, size, size);
 
-        craftMatrix.setInventorySlotContents(0, guiIngredients.get(0));
-        craftMatrix.setInventorySlotContents(1, guiIngredients.get(1));
-        craftMatrix.setInventorySlotContents(2, guiIngredients.get(3));
-        craftMatrix.setInventorySlotContents(3, guiIngredients.get(4));
 
+        if(craftingGUIBuilding.isComplete())
```

Comment:
What about excluding first four as they will be executed every time?

Reply:
On purpose, the index is slightly different